keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -0,0 +1,1088 @@
+returned when updating the policy on an exempt namespace. These checks only consider actual Pod
+resources, not [templated pods].
+
+These checks are also performed when making a dry-run request, which can be an effective way of
+checking for breakages before updating a policy, for example:
+
+```
+kubectl label --dry-run=server --overwrite ns --all <prefix>/enforce=baseline
+```
+
+<<[UNRESOLVED]>>
+
+_Non-blocking: can be decided on the implementing PR_
+
+- What should the timeout be for pod update warnings?
+  - Total is a parameter on the context (query parameter for webhooks). Cap should be
+    `min(timeout_param, hard_cap)`, where the `hard_cap` is a small number of seconds.
+  - Expect evaluation to be fast, so even 3k pods should come in well under the timeout.
+- What should the pod limit be set to?
+  - 3,000 is the
+    [documented](https://github.com/kubernetes/community/blob/master/sig-scalability/configs-and-limits/thresholds.md)
+    scalability limit for per-namespace pod count.
+  - Warnings should be aggregated for large namespaces (soft cap number of warnings, hard cap number
+    of evaluations).
+
+<<[/UNRESOLVED]>>
+
+
+### Admission Configuration
+
+A number of options can be statically configured through the [Admission Configuration file][]:
+
+```
+apiVersion: apiserver.config.k8s.io/v1
+kind: AdmissionConfiguration
+plugins:
+- name: PSPReplacement # Placeholder
+  configuration:
+    defaults:  # Defaults applied when a mode label is not set.
+      enforce:         <default enforce policy level>
+      enforce-version: <default enforce policy version>
+      audit:         <default audit policy level>
+      audit-version: <default audit policy version>
+      warn:          <default warn policy level>
+      warn-version:  <default warn policy version>
+    exemptions:
+      usernames:         [ <array of authenticated usernames to exempt> ]
+      runtimeClassNames: [ <array of runtime class names to exempt> ]
+      namespaces:        [ <array of namespaces to exempt> ]
+...
+```
+
+[Admission Configuration file]: https://github.com/kubernetes/kubernetes/blob/3d6026499b674020b4f8eec11f0b8a860a330d8a/staging/src/k8s.io/apiserver/pkg/apis/apiserver/v1/types.go#L27
+
+#### Defaulting
+
+The default policy level and version for each mode (when no label is present) can be statically
+configured. The default for the static configuration is `privileged` and `latest`.
+
+#### Exemptions
+
+Policy exemptions can be statically configured. Exemptions must be explicitly enumerated, and don’t
+support indirection such as label or group selectors. Requests meeting criteria are ignored by the
+admission controller (enforce, audit and warn) except for recording an [audit
+annotation](#audit-annotations). Exemption dimensions include:
+
+- Usernames: requests from users with an exempt authenticated (or impersonated) username are ignored.
```

Comment:
Highlight use cases in documentation.

Reply:
One implication of this is that exemptions become essentially un-auditable by scraping the API server. There is no breadcrumb on this resource saying that it was exempted and why.

Reply:
the fact that exemptions occurred will show up in metrics, and the reason and particular resource exempted will appear in audit logs